azure-iot-device/azure/iot/device/common/transport/mqtt/mqtt_provider.py

```diff
@@ -217,14 +209,48 @@ def publish(self, topic, payload, qos=1, callback=None):
         """
         logger.info("sending")
         message_info = self._mqtt_client.publish(topic=topic, payload=payload, qos=qos)
-        self._set_operation_callback(message_info.mid, callback)
+        self._op_manager.establish_operation(message_info.mid, callback)
+
+
+class OperationManager(object):
+    """Tracks pending operations and thier associated callbacks until completion.
+    """
+
+    def __init__(self):
+        # Maps mid->callback for operations where a request has been sent
+        # but the reponse has not yet been received
+        self._pending_operation_callbacks = {}
+
+        # Maps mid->mid for responses received that are NOT established in the _pending_operation_callbacks dict.
+        # Necessary because sometimes an operation will complete with a response before the
+        # Paho call returns.
+        # TODO: make this map mid to something more useful (result code?)
+        self._unknown_operation_completions = {}
+
+        self._lock = threading.Lock()
+
+    def establish_operation(self, mid, callback=None):
+        """Establish a pending operation identified by MID, and store its completion callback.
 
-    def _set_operation_callback(self, mid, callback):
-        if mid in self._unknown_operation_responses:
-            # If response already came back, trigger the callback
+        If the operation has already been completed, the callback will be triggered.
+        """
+        # Manually acquire the lock instead of using a handler so it can be released
+        # at different times in the logical branch
```

Comment:
This is OK, but what happens if there's an exception after you acquire and before you release?  It's not terribly likely, but there's no way to recover if it does happen.  I would recommend using a try/finally or a context manager and add a variable to pull the callback out.

like:
do_callback=False
with(lock):
  if blah:
    do_callback = True
if do_callback:
  callback()

I don't like having to add an the extra variable, but having a guarantee that the lock gets released is golden.

Reply:
Plus, my fragile little mind doesn't like having to trace all the logic paths just to make sure it always gets released.  Better to let the language do the work. You know, context managers good, blah blah blah.  

---
In reply to: [287356662](https://github.com/Azure/azure-iot-sdk-python-preview/pull/92#discussion_r287356662) [](ancestors = 287356662)

Reply:
This is a good point - I didn't think about exceptions, since this code clearly cannot raise exceptions between the acquire and release of the lock, and it would be fairly trivial to prove it.

But that could always change, and thus the best practice is definitely to use the context manager. 

---
In reply to: [287356662](https://github.com/Azure/azure-iot-sdk-python-preview/pull/92#discussion_r287356662) [](ancestors = 287356662)